Client/Frontend/UserContent/UserScripts/ReadyState.js

```diff
@@ -0,0 +1,73 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+(function() {
+  // Listen for document ready state
+  document.addEventListener('readystatechange', (event) => {
+    window.webkit.messageHandlers.$<handler>.postMessage({
+      "securitytoken": "$<security_token>",
+      "state": document.readyState
+    });
+  });
+  
+  // Listen for document load state
+  window.addEventListener('load', (event) => {
+    window.webkit.messageHandlers.$<handler>.postMessage({
+      "securitytoken": "$<security_token>",
+      "state": "loaded"
+    });
+  });
+  
+  // Listen for history popped
+  window.addEventListener('popstate', (event) => {
+    if (event.state) {
+      // Run on the browser's next run-loop
+      setTimeout(() => {
+        window.webkit.messageHandlers.$<handler>.postMessage({
+          "securitytoken": "$<security_token>",
+          "state": "popstate"
+        });
+      }, 0);
+    }
+  });
+  
+  // Listen for history pushed
+  const pushState = History.prototype.pushState;
+  History.prototype.pushState = function(state, unused, url) {
```

Comment:
what happens if a webpage redefines `pushState/replaceState`? i guess this feature would just not work?

Reply:
Correct it would just do nothing. I also tested this by doing: `History.prototype.pushState = function() { console.log("Hello"); }` in the Safari Debug Console.

It replaces my script's methods with the one that prints hello, and just breaks the page as well as doing nothing with regards to the script.